### PR TITLE
GH action CI - Benchmark - MOD-4329

### PR DESCRIPTION
**Pull request by GuyAv46**
Also include some general improvements to the GH actions.
After merging this PR to the main branch, we should disable Circle on the VecSim repo

### CODE REVIEW

File: .github/workflows/default.yml [hunk 1]
```diff
@@ -1,17 +1,19 @@
...
-    - cron: "0 0 * * *"
+  workflow_call:
 
 jobs:
   build:
     name: default build job
     runs-on: ubuntu-latest
+    concurrency:
+      group: ${{ github.workflow }}-${{ github.ref }}
+      cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply .github/workflows/default.yml:16]: we enabled it lately on our old CI. I can remove it

File: .github/workflows/benchmark.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                          --github_org ${{ github.repository_owner }} \
+                          --github_branch ${{ github.ref }} \
+                          --benchmark-result-file ${bm_class}_results.json
+            done
+
+  stop-runner:
+    name: Stop self-hosted EC2 runner
+    needs:
+      - start-runner # required to get output from the start-runner job
+      - benchmark # required to wait when the main job is done
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply .github/workflows/benchmark.yml:77]: We always want to take down the runner machine in the end

File: .github/workflows/arm.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
+name: arm flow
+
+on: workflow_call # TODO: add pull_request-labeled with "arm" label when specifying a label is supported
+
+jobs:
+  start-runner:
+    name: Start self-hosted EC2 runner
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'workflow_call' }} || github.event.label.name == 'benchmark'
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply .github/workflows/arm.yml:9]: actually redundant. removing

File: .github/workflows/benchmark.yml [hunk 4]
```diff
@@ -41,34 +41,53 @@ jobs:
...
+      - name: stress test
+        run: |
+          sudo apt install stress-ng -qqy
+          uptime
+          stress-ng -c 1 --timeout 60s --metrics-brief
+          uptime
+          stress-ng --stream 1 -t 60 --metrics-brief
+          uptime
+          stress-ng --ipsec-mb=1 -t 60 --metrics-brief
+          uptime
```
Comment (DvirDukhan) [inline .github/workflows/benchmark.yml:68]: do we still need this?

File: .github/workflows/focal.yml [hunk 5]
```diff
@@ -5,6 +5,8 @@ on: workflow_call # TODO: add pull_request-labeled with "focal" label when speci
 jobs:
   focal:
     runs-on: ubuntu-latest
+    container:
+      image: ubuntu:focal
```
Comment (DvirDukhan) [inline .github/workflows/focal.yml:9]: does it works with `sudo`?

File: src/VecSim/spaces/spaces.cpp [hunk 6]
```diff
@@ -53,25 +53,29 @@ CalculationGuideline FP64_GetCalculationGuideline(size_t dim) {
 
 void SetDistFunc(VecSimMetric metric, size_t dim, dist_func_t<float> *index_dist_func) {
 
+    static const Arch_Optimization arch_opt = getArchitectureOptimization();
```
Comment (DvirDukhan) [inline src/VecSim/spaces/spaces.cpp:56]: will this work for multiple indexes?

File: .github/workflows/benchmark.yml (hunk 4 above)
Comment (GuyAv46) [reply .github/workflows/benchmark.yml:68]: Yes, Filipe asked me to keep this and he will add a time series with this data

File: src/VecSim/spaces/spaces.cpp (hunk 6 above)
Comment (GuyAv46) [reply src/VecSim/spaces/spaces.cpp:56]: yes. the function will be called only once (for each type) and then will use the `arch_opt` as a constant

File: .github/workflows/focal.yml (hunk 5 above)
Comment (GuyAv46) [reply .github/workflows/focal.yml:9]: No. fixed

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**